### PR TITLE
Support xml-stylesheet processing instructions

### PR DESCRIPTION
**Pull request by thewilkybarkid**
# ↪️ Pull Request

Adds `@parcel/transformer-xml`, starting with support for the `xml-stylesheet` processing instruction.

Relates to #2901.

## 💻 Examples

A HTML file containing:

```html
<object data="image.svg"></object>
```

`image.svg`:

```svg
<?xml-stylesheet href="style.css"?>
<svg xmlns="http://www.w3.org/2000/svg">
  <rect width="100" height="100"/>
</svg>
```

`style.css`:

```css
:root {
  fill: red;
}
```

The square should appear red rather than black.

## 🚨 Test instructions

(I've added an integration test, but unable to get it working locally.)

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by thewilkybarkid** [comment]
There's more that can be added (`xlink:href` is the obvious one), but some SVG specifics too in a complementary SVG transformer (more details in #2901).

File: yarn.lock [hunk 1]
```diff
@@ -8032,6 +8150,39 @@ jsdom@^15.2.1:
...
+    symbol-tree "^3.2.4"
+    tough-cookie "^4.0.0"
+    w3c-hr-time "^1.0.2"
+    w3c-xmlserializer "^2.0.0"
+    webidl-conversions "^6.1.0"
+    whatwg-encoding "^1.0.5"
+    whatwg-mimetype "^2.3.0"
+    whatwg-url "^8.5.0"
+    ws "^7.4.5"
+    xml-name-validator "^3.0.0"
```
Comment (thewilkybarkid) [inline yarn.lock:8184]: This is the only local change, not sure where the other changes have come from.

**Comment by thewilkybarkid** [comment]
How do I get the integration test to pick up the new transformer? Obviously, it's not published (yet), but I assumed these would use the local versions?

**Comment by mischnic** [comment]
You need to rename "packages/transformers/xml/src/index.js" to "XMLTransformer.js"

**Comment by mischnic** [comment]
I'm not entirely sure whether a single transformer for atom, rss, svg and xml is the right approach.

And another problem with your changes is that `import x from "./icon.svg";` is allowed, while it should require `url:./icon.svg` like other images (as it is now).

**Comment by thewilkybarkid** [comment]
I imagined having a SVG transformer to use in combination with this, so this handles generic XML features and the other the SVG format. I included Atom/RSS since they're common extensions using XML, but they could have their own transformers too if needed.

Ok to change the format (though I'll be changing it back locally).

**Comment by thewilkybarkid** [comment]
Thinking about, XLink isn't so generic: `<svg:a>` should behave differently to `<svg:image>` (the former should have a stable path).

**Comment by thewilkybarkid** [comment]
Given https://github.com/parcel-bundler/parcel/pull/6565 should I remove the `url:` prefix?

File: packages/transformers/xml/src/XMLTransformer.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+export default (new Transformer({
+  canReuseAST({ast}) {
+    return ast.type === 'jsdom' && semver.satisfies(ast.version, '^16.6.0');
+  },
+
+  async parse({asset}) {
+    return {
+      type: 'jsdom',
+      version: '16.6.0',
+      program: new JSDOM(await asset.getBuffer(), {
```
Comment (devongovett) [inline packages/transformers/xml/src/XMLTransformer.js:17]: JSDOM seems a bit heavyweight for this use case. Should we use a simpler XML parser for this instead?

File: packages/core/integration-tests/test/integration/xml-stylesheet-processing-instruction/styled.xml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?xml-stylesheet href="style1.css"?>
+<?xml-stylesheet href="style2.css?>
+<?xml-smilesheet href="style2.css"?>
```
Comment (devongovett) [inline packages/core/integration-tests/test/integration/xml-stylesheet-processing-instruction/styled.xml:3]: Is that intentional?

File: packages/configs/default/index.json [hunk 4]
```diff
@@ -40,6 +40,7 @@
     "style:*.vue": ["@parcel/transformer-vue"],
     "custom:*.vue": ["@parcel/transformer-vue"],
     "*.{png,jpg,jpeg,webp}": ["@parcel/transformer-image"],
+    "*.{atom,rss,svg,xml}": ["@parcel/transformer-xml"],
```
Comment (devongovett) [inline packages/configs/default/index.json:43]: Seems like we should probably have different transformers for each of these. Atom, RSS, and SVG will all have different tags and attributes that we'll need to handle to create dependencies for sub resources. Maybe keep this transformer specific to `.xml` files?

**Comment by devongovett** [comment]
Curious what your use case is for this? How common is it to have a `.xml` file with a stylesheet attached? I've never seen that before.

**Comment by thewilkybarkid** [comment]
It's how Graphviz attaches stylesheets (refs #6544).

**Comment by thewilkybarkid** [comment]
Probably better to just make this an SVG transformer (as the name's now free). It might be that an SVG using a PI containing (S)CSS (as opposed to others like XSLT) should become a `<style>@import(...)</style>` so it can be inlined and then optimised. (External files aren't loaded when using an ``.)

If there are future uses for processing XML-based formats then it could be revisited, but the expected behaviour might be different anyway.

**Comment by thewilkybarkid** [comment]
Closing in favour of #6595.

**Pull request closed without merging**